File: backend/feature/label/src/application/use_case/get_all_labels_use_case.rs
use std::sync::Arc;

use async_trait::async_trait;

use crate::{
    application::{error::label_error::LabelError, gateway::label_repository::LabelRepository},
    domain::entity::label::Label,
};

#[async_trait]
pub trait GetAllLabelsUseCase: Send + Sync {
    async fn execute(&self) -> Result<Vec<Label>, LabelError>;
}

pub struct GetAllLabelsUseCaseImpl {
    label_repository: Arc<dyn LabelRepository>,
}

impl GetAllLabelsUseCaseImpl {
    pub fn new(label_repository: Arc<dyn LabelRepository>) -> Self {
        Self { label_repository }
    }
}

#[async_trait]
impl GetAllLabelsUseCase for GetAllLabelsUseCaseImpl {
    async fn execute(&self) -> Result<Vec<Label>, LabelError> {
        self.label_repository.get_all_labels().await
    }
}

File: frontend/src/lib/label/application/useCase/getAllLabelsUseCase.ts
import type { LabelRepository } from '$lib/label/application/gateway/labelRepository';
import type { Label } from '$lib/label/domain/entity/label';

export class GetAllLabelsUseCase {
	constructor(private readonly labelRepository: LabelRepository) {}

	execute(): Promise<Label[]> {
		return this.labelRepository.getAllLabels();
	}
}
